DataDefinitions/Cargo.cs

```diff
@@ -154,28 +154,115 @@ private void OnDeserialized(StreamingContext context)
         public Cargo() { }
 
         [JsonConstructor]
-        public Cargo(string edname, int total, int? price = null)
+        public Cargo(string edname)
         {
-            this.commodityDef = CommodityDefinition.FromEDName(edname);
-            this.price = price ?? commodityDef.avgprice;
-            this.total = total;
-            haulageData = new List<Haulage>();
+            commodityDef = CommodityDefinition.FromEDName(edname);
         }
 
         public event PropertyChangedEventHandler PropertyChanged;
-
+        
         public void NotifyPropertyChanged(string propName)
         {
-            this.total = this.stolen + this.haulage + this.owned;
             PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(propName));
         }
 
         public void CalculateNeed()
         {
-            if (this != null && this.haulageData != null && this.haulageData.Any())
+            if (haulageData != null && haulageData.Any())
+            {
+                need = haulageData.Sum(h => h.need);
+            }
+        }
+
+        public void CalculateWeightedPrice(int thisPrice, int amount)
+        {
+            if (amount == 0) { return; }
+            var weightedValueSum = (weightedAvgPrice * total) + (thisPrice * amount);
+            var weightedQtySum = total + amount;
+            weightedAvgPrice = weightedValueSum / weightedQtySum;
+            NotifyPropertyChanged("price");
+        }
+
+        /// <summary> </summary>
+        /// <param name="cargoType">The type of cargo to add</param>
+        /// <param name="amount">The amount of cargo to add</param>
+        /// <param name="cargoHaulageData">Add or update haulage instance</param>
+        /// <param name="_price">The acquisition price per unit (if not zero)</param>
+        public void AddDetailedQty(CargoType cargoType, int amount, Haulage cargoHaulageData = null, int _price = 0)
+        {
+            CalculateWeightedPrice(_price, amount);
+            switch (cargoType)
+            {
+                case CargoType.haulage:
+                    {
+                        haulage += amount;
+                        if (cargoHaulageData != null) 
+                        {
+                            for (int i = 0; i < haulageData.Count; i++)
+                            {
+                                if (haulageData[i].missionid == cargoHaulageData.missionid)
+                                {
+                                    haulageData[i] = cargoHaulageData;
+                                    return;
```

Comment:
A `return` this deep is always a bit of a loaded gun should the code be blindly moved into another context.

I would instead of the manual `for` loop use `haulageData.FindIndex()` with a predicate and update if found or Add() if not.